### PR TITLE
Add useEmulator() to Storage

### PR DESCRIPTION
**Pull request by schmidt-sebastian**
This adds emulator support to the Cloud Storage SDK. Since public facing Storage URLs are in the form gs://bucket/object, they are not affected by this change. Instead, only `firebasestorage.googleapis.com` is changed to the emulator URL.

### CODE REVIEW

File: FirebaseStorage/CHANGELOG.md [hunk 1]
```diff
@@ -1,3 +1,7 @@
+# Unreleased
+- [added] Added `FirebaseDatabase.useEmulator()`, which allows the Storage SDK to
```
Comment (morganchen12) [inline FirebaseStorage/CHANGELOG.md:2]: caught copy-pastin'

File: FirebaseStorage/Sources/FIRStorage.m [hunk 2]
```diff
@@ -299,6 +307,27 @@ - (void)setCallbackQueue:(dispatch_queue_t)callbackQueue {
   _fetcherServiceForApp.callbackQueue = callbackQueue;
 }
 
+- (void)useEmulatorWithHost:(NSString *)host port:(NSInteger)port {
+  if (host.length == 0) {
+    [NSException raise:NSInvalidArgumentException format:@"Cannot connect to nil or empty host."];
+  }
+
+  if (port < 0) {
+    [NSException raise:NSInvalidArgumentException format:@"Port must be greater than zero."];
```
Comment (morganchen12) [inline FirebaseStorage/Sources/FIRStorage.m:316]: nit: greater than or equal to

File: FirebaseStorage/Sources/FIRStorageUtils.h [hunk 3]
```diff
@@ -17,6 +17,7 @@
 #import <Foundation/Foundation.h>
 
 @class FIRStoragePath;
```
Comment (morganchen12) [inline FirebaseStorage/Sources/FIRStorageUtils.h:19]: Can this forward declaration now be removed?

File: FirebaseStorage/Tests/Unit/FIRStorageDeleteTests.m [hunk 4]
```diff
@@ -102,7 +102,29 @@ - (void)testSuccessfulFetch {
...
+                                                                      [expectation fulfill];
+                                                                    }];
+  [task enqueue];
+
+  [FIRStorageTestHelpers waitForExpectation:self];
+}
+
+- (void)testUnsuccessfulDeleteUnauthenticated {
   XCTestExpectation *expectation =
       [self expectationWithDescription:@"testUnsuccessfulFetchUnauthenticated"];
```
Comment (morganchen12) [inline FirebaseStorage/Tests/Unit/FIRStorageDeleteTests.m:129]: nit: change the expectation name too

**Review by morganchen12 — APPROVED** [review]
LGTM, had a few small comments.

File: FirebaseStorage/CHANGELOG.md (hunk 1 above)
Comment (schmidt-sebastian) [reply FirebaseStorage/CHANGELOG.md:2]: Ehm.

File: FirebaseStorage/Sources/FIRStorage.m (hunk 2 above)
Comment (schmidt-sebastian) [reply FirebaseStorage/Sources/FIRStorage.m:316]: Fixed

File: FirebaseStorage/Sources/FIRStorageUtils.h (hunk 3 above)
Comment (schmidt-sebastian) [reply FirebaseStorage/Sources/FIRStorageUtils.h:19]: It's still used by other APIs in this file.

File: FirebaseStorage/Tests/Unit/FIRStorageDeleteTests.m (hunk 4 above)
Comment (schmidt-sebastian) [reply FirebaseStorage/Tests/Unit/FIRStorageDeleteTests.m:129]: I changed the test name back (all the test use Fetch). I wanted to clean this up originally but then gave up.

File: FirebaseStorage/Tests/Integration/FIRStorageIntegrationTests.m [hunk 5]
```diff
@@ -510,10 +511,11 @@ - (void)testGetFile {
                       NSString *data = [NSString stringWithContentsOfURL:fileURL
                                                                 encoding:NSUTF8StringEncoding
                                                                    error:NULL];
-                      NSString *expectedData = @"Hello World";
-                      XCTAssertEqualObjects(data, expectedData);
-                      XCTAssertEqualObjects([snapshot description], @"<State: Success>");
-                      [expectation fulfill];
+                      [ref dataWithMaxSize:128
```
Comment (morganchen12) [inline FirebaseStorage/Tests/Integration/FIRStorageIntegrationTests.m:514]: nit: maybe use `fileData.size` so this test doesn't break if fileData changes

**Review by morganchen12 — APPROVED** [review]

File: FirebaseStorage/Tests/Integration/FIRStorageIntegrationTests.m [hunk 6]
```diff
@@ -117,8 +119,8 @@ - (void)setUp {
     }
 
     for (NSString *emptyFile in emptyFiles) {
-      FIRStorageReference *file = [[FIRStorage storage].reference child:emptyFile];
-      [file putData:[NSData data]
+      FIRStorageReference *file = [self.storage.reference child:emptyFile];
+        [file putData:[@"foo" dataUsingEncoding:NSUTF8StringEncoding]
```
Comment (morganchen12) [inline FirebaseStorage/Tests/Integration/FIRStorageIntegrationTests.m:123]: weird indent?

File: FirebaseStorage/Tests/Integration/FIRStorageIntegrationTests.m [hunk 7]
```diff
@@ -77,6 +77,8 @@ - (void)setUp {
   self.app = [FIRApp defaultApp];
   self.auth = [FIRAuth authWithApp:self.app];
   self.storage = [FIRStorage storageForApp:self.app];
+    
+   [ self.storage useEmulatorWithHost:@"localhost" port:9199];
```
Comment (morganchen12) [inline FirebaseStorage/Tests/Integration/FIRStorageIntegrationTests.m:81]: fix whitespace

**Review by morganchen12 — APPROVED** [review]

File: FirebaseStorage/Tests/Integration/FIRStorageIntegrationTests.m (hunk 6 above)
Comment (schmidt-sebastian) [reply FirebaseStorage/Tests/Integration/FIRStorageIntegrationTests.m:123]: This is just for Abe to do some debugging

File: FirebaseStorage/Tests/Integration/FIRStorageIntegrationTests.m (hunk 5 above)
Comment (schmidt-sebastian) [reply FirebaseStorage/Tests/Integration/FIRStorageIntegrationTests.m]: Fixed

**Comment by paulb777** [comment]
Will you update the integration tests in CI to use the emulator either in this PR or a subsequent one?

**Comment by schmidt-sebastian** [comment]
> Will you update the integration tests in CI to use the emulator either in this PR or a subsequent one?

There are still some issues in the Emulator that @abeisgoat is working on fixing. We can migrate the tests afterwards.

**Pull request merged**